Yosemite/YosemiteTests/Stores/ShipmentStoreTests.swift

```diff
@@ -202,6 +216,148 @@ class ShipmentStoreTests: XCTestCase {
+
+            let group1 = self.viewStorage.loadShipmentTrackingProviderGroup(siteID: self.sampleSiteID, providerGroupName: self.sampleCountryName)
+
+            let groupProviders = group1?.providers
+
+            XCTAssertNotEqual(groupProviders?.count, 0)
+
+            XCTAssertNotNil(group1)
+            XCTAssertEqual(group1?.toReadOnly(), self.australia())
+
+            expectation.fulfill()
+        }
+
+        shipmentStore.onAction(action)
+        wait(for: [expectation], timeout: Constants.expectationTimeout)
+    }
+
+    /// Verifies that `ShipmentAction.synchronizeShipmentTrackingProviders` returns an error whenever there is an error response from the backend.
+    ///
+    func testRetrieveShipmentTrackingGroupListReturnsErrorUponReponseError() {
+        let expectation = self.expectation(description: "Retrieve shipment tracking provider group list error response")
+        let shipmentStore = ShipmentStore(dispatcher: dispatcher, storageManager: storageManager, network: network)
+
+        network.simulateResponse(requestUrlSuffix: "orders/" + String(sampleOrderID) + "/" + "shipment-trackings/providers",
+                                 filename: "shipment_tracking_plugin_not_active")
+        let action = ShipmentAction.synchronizeShipmentTrackingProviders(siteID: sampleSiteID, orderID: sampleOrderID) { error in
+            XCTAssertNotNil(error)
+            expectation.fulfill()
+        }
+
+        shipmentStore.onAction(action)
+        wait(for: [expectation], timeout: Constants.expectationTimeout)
+    }
+
+    /// Verifies that `ShipmentAction.synchronizeShipmentTrackingProviders` returns an error whenever there is no backend response.
+    ///
+    func testRetrieveShipmentTrackingGroupListReturnsErrorUponEmptyResponse() {
+        let expectation = self.expectation(description: "Retrieve shipment tracking provider grup list empty response")
+        let shipmentStore = ShipmentStore(dispatcher: dispatcher, storageManager: storageManager, network: network)
+
+        let action = ShipmentAction.synchronizeShipmentTrackingProviders(siteID: sampleSiteID, orderID: sampleOrderID) { error in
+            XCTAssertNotNil(error)
+            expectation.fulfill()
+        }
+
+        shipmentStore.onAction(action)
+        wait(for: [expectation], timeout: Constants.expectationTimeout)
+    }
+
+    /// Verifies that `upsertShipmentTrackingProviderData` does not produce duplicate entries.
+    ///
+    func testUpdateRetrieveShipmentTrackingProviderGroupListEffectivelyUpdatesPreexistantShipmentTrackingData() {
+        let shipmentStore = ShipmentStore(dispatcher: dispatcher, storageManager: storageManager, network: network)
+        XCTAssertEqual(self.viewStorage.countObjects(ofType: Storage.ShipmentTrackingProviderGroup.self), 0)
+
+        let group = DispatchGroup()
+
+        group.enter()
+        shipmentStore.upsertTrackingProviderDataInBackground(siteID: sampleSiteID,
```

Comment:
This write to Core Data should probably happen outside of the `DispatchGroup()` — I was seeing a race condition occur while debugging.

See my test branch → https://github.com/woocommerce/woocommerce-ios/blob/try/689-mark3-bummy/Yosemite/YosemiteTests/Stores/ShipmentStoreTests.swift#L285